### PR TITLE
feat: split out go module defs into package

### PR DESCRIPTION
**Pull request by jedevc**
Finally! This patch splits up the codegen for go modules to produce the main external struct defs into a separate sub-package. There are a couple of reasons for doing this:
- Tighter control over exposed structs and their internal private methods
- Ability to import the generated structs from a sub-package

Before:

```
❯ tree .
.
├── api.go
├── dagger.gen.go
├── dagger.json
├── go.mod
├── go.sum
├── main.go
├── querybuilder
│   ├── marshal.go
│   └── querybuilder.go
└── README.md

2 directories, 9 files
```

After:

```
❯ tree .
.
├── api.go
├── dagger
│   └── dagger.gen.go
├── dagger.gen.go
├── dagger.json
├── go.mod
├── go.sum
├── main.go
├── querybuilder
│   ├── marshal.go
│   └── querybuilder.go
└── README.md

3 directories, 10 files
```

Essentially, we split up `dagger.gen.go` into a top-level `dagger.gen.go`, and a sub-package `dagger/dagger.gen.go`.

- Top-level `dagger.gen.go` contains type-aliases to all the public sub-package types, any required interfacey/struct method implementations, and the implementation of `main` and `invoke`.
- Packaged `dagger/dagger.gen.go` contains all the go-generated templates for the served schema types.
	- This package can now also be imported by sibling packages, see the new test `TestModuleGoUtilsPkg`, which tests for #6661.

---

Initially this might sound simple - and prior to the addition of interfaces, it was, however, these add a lot of confusion. Specifically, interface codegen is added alongside the main code, and includes adding new methods on user-defined structs as well as creating an implementation of user-defined interfaces.

So where should these new definitions belong? They actually have to go at the same level as the definition of the struct/interface, so that 1. we can actually attach the methods, and 2. so we can actually access the interface in arg/return types (we can't just copy the interface into the sub-package, since it might also refer to user-defined concrete types, which we can't copy).

But this results in a couple of complications:
- We defined a LoadFromID method for our own interfaces - we can no longer attach these to the client, since the client is now part of the sub-package. This is easy to resolve, we just make this a top-level method, it doesn't actually *need* to be on the client object, that was just a convention.
- We have to manually construct core types from our returns - for this, we need access to the `q`/`c` values on the core types. We just rename these and make them public, calling them `Query`/`Client` respectively. This is *annoying* that we have to do this at all (but there are also legitimate use cases for accessing these directly); as a non-urgent follow-up we should rework these structs so they *can't* be misused, and those fields can't be manipulated inappropriately.

TODOs:
- [x] Make sure that the non-zenith codegen is unaffected
- [x] It doesn't seem like the interfaces loadFromID methods are getting served back to themselves (I left an XXX here)
- [x] Ensure interface code docs in `module_interfaces.go` are up-to-date
- [x] Tests are failing because of removed "code-generate foreign type"
- [x] Write a sub-package util import test
- [ ] ~~See if we can avoid exposing `Client`/`Query` everywhere~~ (not convinced we need to do this right now)

### CODE REVIEW

File: cmd/codegen/generator/go/templates/module_interfaces.go [hunk 1]
```diff
@@ -570,11 +571,17 @@ func (spec *parsedIfaceType) concreteMethodExecuteQueryCode(method *funcTypeSpec
...
+				d := Dict{
+					Id("Query"):  Id("querybuilder").Dot("Query").Call().Dot("Select").Call(Lit(loadFromIDQueryName)).Dot("Arg").Call(Lit("id"), Id("id")),
+					Id("Client"): Id("r").Dot("Client"),
+				}
+				if strings.Contains(loadFromIDQueryName, "IfaceFromID") {
+					// XXX: do a better hack here
+					// ...but also, this *clearly* isn't right - it feels like
+					// schemas aren't being served back
+					d[Id("id")] = Op("&").Id("id")
+				}
```
Comment (jedevc) [inline cmd/codegen/generator/go/templates/module_interfaces.go:583]: So if I remove this (essentially the `id` would be fetched later, right), I get this bizarre error in `TestModuleIfaceBasic`:

```
    multi.go:85: 27: [41.4s]         	            	
    multi.go:85: 27: [41.4s]         	            	error: parse selections: parse field "loadOtherIfaceFromID": Query has no such field: "loadOtherIfaceFromID"
    multi.go:85: 27: [41.4s]         	Test:       	TestIface/dynamicOtherIfaceByIfaceList
```

So I thought that https://github.com/dagger/dagger/pull/6219 would serve it's own bindings to itself? It seems like it's not, and something is wrong here.

File: sdk/go/client.go [hunk 2]
```diff
@@ -75,9 +75,9 @@ func Connect(ctx context.Context, opts ...ClientOpt) (*Client, error) {
 	gql := errorWrappedClient{graphql.NewClient("http://"+conn.Host()+"/query", conn)}
 
 	c := &Client{
-		c:    gql,
-		conn: conn,
-		q:    querybuilder.Query(),
+		Client: gql,
+		Query:  querybuilder.Query(),
+		conn:   conn,
 	}
```
Comment (jedevc) [inline sdk/go/client.go:81]: So potentially this is controversial, so wanted to flag it :white_flag:

This changes these properties (and on all the core types) from being private to being public - this isn't *exactly* necessary for the non-module SDK, but without this, we end up having to maintain multiple variants of the codegen, one with public fields (for modules) and one with private fields (for the general case).

Personally?
- I'd rather have the module API and today's API line up as *much* as possible - even if that requires changing today's API to get there.
- I dislike needing to have these fields public at all (access to them is fine, but ideally we shouldn't be able to write to them *after* initialization). I can't think of a clever way to do this though really, but some ideas for completeness:
	- We could add new `NewDirectory`/etc calls that set private properties - but while these are "public" fields, they're not really meant for users to call.
	- We could have all of our generated types implement a custom `Setup` interface, which could then allow setting the query/client fields exactly once. We could hide the method implementation, and then just have a top-level function `Setup` (this is a variant of the above, that doesn't require a new public function for each type, just one `Setup`).
	- We could "freeze" and copy `Client` to `c` and `Query` to `q` on first use of it, then changes to the public fields don't affect private fields (so still allows initialization). 

Kinda not sure, so curious to hear if there are any issues with this. cc @vito @sipsma

**Comment by sipsma** [comment]
> * Packaged dagger/dagger.gen.go contains all the go-generated templates for the served schema types.

Tried locally and it looks like installed deps only end up in the top-level dagger.gen.go, not in `dagger/dagger.gen.go`, which seems unexpected.

> * Top-level dagger.gen.go contains type-aliases to all the public sub-package types, any required interfacey/struct method implementations, and the implementation of main and invoke.

Is this up to date? I don't see any type aliases being used in the top-level `dagger.gen.go`

**Comment by jedevc** [comment]
> Tried locally and it looks like installed deps only end up in the top-level dagger.gen.go, not in `dagger/dagger.gen.go`, which seems unexpected.

Hm, I'm not seeing that. I put together a git repo as an example: https://github.com/jedevc/dagger-module-split-gen-example

I've installed `github.com/kpenfound/dagger-modules/golang`: https://github.com/jedevc/dagger-module-split-gen-example/blob/main/dagger.json#L4-L9

This appears in the sub-package `dagger.gen.go`: https://github.com/jedevc/dagger-module-split-gen-example/blob/main/dagger/dagger/dagger.gen.go#L3568-L3576

> > * Top-level dagger.gen.go contains type-aliases to all the public sub-package types, any required interfacey/struct method implementations, and the implementation of main and invoke.
> 
> Is this up to date? I don't see any type aliases being used in the top-level `dagger.gen.go`

The type aliases definitely look like they're there: https://github.com/jedevc/dagger-module-split-gen-example/blob/main/dagger/dagger.gen.go#L64-L65

The current behavior is expected to me, could you clarify your comments sorry!

**Comment by sipsma** [comment]
> The current behavior is expected to me, could you clarify your comments sorry!

No I think this must be my bad, when running locally I did see the new `dagger/dagger.gen.go` created, but looking again at the `dagger.json` has version `v0.9.9` 🤔 I'm guessing I somehow accidentally ran the `init` with this PR but then switched to v0.9.9 somehow for the `install` 🤷‍♂️ Sorry I was trying to review this quickly before I had to run off to do other things, must have used the wrong terminal or something at some point.

Either way, ran again and looks like what you described. Will finish reviewing today!

**Review by sipsma — APPROVED** [review]
LGTM, the tradeoffs of detaching Load*FromID for interfaces and making Query/Client public make sense to me 👍

I do remember discussing recently the idea of no longer trying to directly execute the user code but instead just copy it to the generated package and attach all the needed types there alongside the `main` func, etc. Then we'd just have the bare minimum type aliases in the "top-level" `dagger.gen.go` to support IDE autocomplete.

I do think that would probably result in a cleaner end state and centralize as much as possible in the generated package, but that's ultimately not really a user-facing benefit, more just simpler for dagger devs to reason about in the future. So I think this PR is a good stepping stone towards that, not worth blocking on all that now.

**Pull request merged**